python/docs/dlis/investigation.rst

```diff
@@ -0,0 +1,171 @@
+ 4. Can build dlisio_ and lfp_ from source code and have enough knowledge to
+    debug it
+
+
+Overview
+--------
+
+When error happens on load, it usually means that file structure itself is
+spoiled. So you should be able to find error just by checking VR and LR
+headers.
+
+When the file itself loads, it means that data layout is fine, which
+*probably* means that *data* itself is fine and is written as intended. Errors
+on accessing objects indicate that intention didn't match RP66 specification.
+
+
+Visual inspection
+-----------------
+
+Visually inspect the file to decide if failure cause is obvious.
+
+Open the file in binary editor, scroll through it.
+
+* Make sure that it looks like every other dlis file that opens correctly.
+  Regardless of the extension, file contents might not be dlis.
+* See if you notice zeroed-out bytes at the end of the file. This likely means
+  there was an error during file creation and data is truncated.
+
+Find failure offset
+-------------------
+
+.. note::  Remember that dlisio fails when it definitely can't process the
+           data anymore. Which means that real error might have happened
+           somewhere **before** the failure point indicates.
+
+Error on load
+^^^^^^^^^^^^^
+If error happened during data load, you might see in the error message last
+known absolute offset aka physical tell (decimal). Usually offset would point
+to the position of LRS which dlisio was unable to process.
+
+Error on parse
+^^^^^^^^^^^^^^
+If error happened during parsing, you'll likely find object set, object or
+attribute identification in the error message. You may search the binary
+file in UTF-8 mode to find these markers in the respective logical file,
+thus establishing some resemblance of failure offset.
+
+Failure offset hint not present
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In some cases this might not be enough and you'll have to dig into the code,
+which won't be easy and straightforward.
+
+Some first steps:
+
+* Build dlisio_ and lfp_ in debug mode.
+* Start Python debugging process, find failing C++ call, start gdb (or
+  similar C++ debugger) and connect to Python process.
+* While debugging further keep in mind that dlisio uses layered-file-protocol
+  library. It means that all the offsets passed inside dlisio are not absolute
+  (TIF or rp66 header bytes are not included), so you'll have to dig into lfp_
+  library to find out what really happens.
+* You might need to modify the code to give yourself more anchor points if
+  failure, for example, happens late in the long loop.
+
+
+Figure out failure reason
+-------------------------
+
+Results of your investigation might be something that makes sense - certain
+specification violation, missing bytes or some specific but consistent between
+several files error.
+
+Sometimes, however, it won't be possible to figure out what exactly went wrong.
+Then failure can't be classified and there is little that can be fixed.
```

Comment:
5 lines to tell you the obvious thing: "if you can't find the cause, you can't fix it"? Would remove all together

Reply:
The condense it a bit? Something like this should capture it?

> We've seen files where there is a lot of incorrectly written bytes and no obvious pattern to it. In such cases, the effort of fixing them drastically increases